WIP - Dev work for #9

krushiraj opened the pull request:
Started the work for #9 . This is currently in Work in progress status.

krushiraj committed:
WIP - Inital commit for #9

krushiraj committed:
DataListContainer added

* DataListContainer:
  --A generic container that accepts the list of items and iterates over it to render a list. It also provides pagination options.
  --childComponent prop will accept a component class that will be rendered if  the items in the list doesn't contain any property called `child`, which holds the component class.
  --Added QuestionListItem component to test.

TODO: DataFilter which passes filtered data to the above component. DataDetails which will show the detailed view of the list item selected.

krushiraj commented:
@pbteja1998 I have added a generic component to component to render the list of data provided. Have a look at that [here](https://openrank-frontend-git-dev-issue-9.coderplex.now.sh/dashboard). If this feels okay, then I can proceed with the configurations and improvements on the same. Meanwhile, I'll be working on filter and preview pane(right side one).

pbteja1998 commented on src/components/DataDetails.vue:
```diff
@@ -0,0 +1,27 @@
+
+<script>
+export default {
+  name: 'Card',
```
Don't name this as Card. There is already a component with name `Card`. Let us not give same name to different components.

krushiraj committed:
Update DataDetails.vue

Changed name, previous bane was not interfere for this. A left over work from copy paste

krushiraj committed:
Update DataFilter.vue

Changed name, previous bane was not interfere for this. A left over work from copy paste

pbteja1998 commented on src/views/Dashboard.vue:
```diff
@@ -0,0 +1,23 @@
+    <v-col cols="3">
+
+    </v-col>
+    <v-col cols="6">
```
Dashboard is not responsive. It does not look good in mobile. Please fix this. 

There are two ways in which you can go about this. 

1. Just add `xs="12"` and `md="6"` to `v-col`. Similarly for left and right panes. So that in mobile all panes appear below each other.

1. Or Try to make left and right pane collapse to side in mobile view. There should be hamburger icon or something similar on top which should be used to make the sidebars visible.

If you think 2nd implementation is difficult, go ahead with first one.

krushiraj replied:
2nd one was the thought. I wanted to make sure if you're okay with this generic list renderer. This commit is just to give you a glance. Will have the required changes by EOD.

pbteja1998 replied:
Oh..Ok. The generic list looks very good. 

Let us take a use-case.

For example, we are rendering list of questions, Assume that each question will have `title`, `category`, `difficulty_level`, `list_of_tags`. 

Now tell me what would you pass to this generic component and in what structure do you need to pass.

krushiraj replied:
The component I wrote expects a child component and props that are to be passed.

pbteja1998 replied:
What I wanted to ask was - is there some processing that needs to be done from the response that we get from api. Or we can directly pass the api response to the component and it will take care of rest.

For example, we might get a list of question objects as an api response. Now we have to preprocess that list( everytime we want to use)  to fit the structure that you are using.

So is it possible to remove this extra step?

krushiraj replied:
Essentially this should remove that step as the data that is received by this component is forwarded to child component. This component will have no processing steps on data. It only iterates and gives a data item from list to child component. 

If at all any processing is needed then it can be done in the child component that we pass a prop.

pbteja1998 replied:
We don't want to do any processing at any component.
I will send the data like the following.
```
{
    fields: ["id", "list_of_tags", "category", "difficulty_level"]
    data: [
        ....
        {
            id: 11,
            list_of_tags: ["a", "b", ...],
            category: "dp",
            difficulty_level: "easy"
        },
        ....
    ],
}
```

Now, the component should be able to render this. I am asking if something like this is possible.

krushiraj replied:
Yes this is possible.

pbteja1998 commented on src/router/index.js:
```diff
@@ -39,5 +40,10 @@ export default new Router({
+    {
+      path: '/dashboard',
+      name: 'dashboard',
+      component: Dashboard,
```
Add dashboard to the logged in view of the header.

krushiraj replied:
Okay. Will do that once I'm done with all the three components.

pbteja1998 commented on src/components/DataListContainer.vue:
```diff
@@ -0,0 +1,110 @@
+<template>
+  <v-container fluid>
```
adding `fluid` did not make any change in the UI. Check once and remove it if it is not making any difference.

pbteja1998 commented on src/components/DataListContainer.vue:
```diff
@@ -0,0 +1,110 @@
+
+export default {
+  name: 'DataListContainer',
+  data: () => ({
```
Let us remove this and add these as props.

pbteja1998 commented on src/views/Dashboard.vue:
```diff
@@ -0,0 +1,23 @@
+
+    </v-col>
+    <v-col cols="6">
+      <DataListContainer />
```
Pass the required props here.

pbteja1998 commented on src/components/DataListContainer.vue:
```diff
@@ -0,0 +1,110 @@
+  methods: {
+    getChildComponent(child) {
+      return child || this.childComponent;
+    },
```
We want this component to be as dumb as it can. If at all it is possible to remove this, then do it.

krushiraj replied:
> @pbteja1998 `childComponent` expects a component which is used to render the list items. `dummyData` contains the props that are expected by the `childComponent`. In addition, you may want some children to be of different component then you can add something like this:
> 
> ```js
> //Assume that card is imported already.
> const dummyData = [
> ...,
> {
>     child: Card,
>     props: {
>       title: 'Frozen Yogurt',
>       category: 'Coding',
>       tags: ['some', 'random', 'tag'],
>       //props that are expected by the Card component.
>     },
>   },
> ...,
> ```

This method is required for over-riding the child component.

pbteja1998 commented on src/views/Dashboard.vue:
```diff
@@ -0,0 +1,23 @@
+
+    </v-col>
+  </v-row>
+</template>
```
As we will be using similar structure of `v-row` and `v-col` for many pages, extract this into a separate component and add slots for left, right and middle panes.

krushiraj replied:
Yeah sure

krushiraj commented on src/components/DataListContainer.vue:
```diff
@@ -1,10 +1,110 @@
+    childComponent: QuestionListItem,
+    itemsPerPageOptions: [4, 8, 12],
+    itemsPerPage: 4,
+    items: dummyData,
```
@pbteja1998 `childComponent` expects a component which is used to render the list items. `dummyData` contains the props that are expected by the `childComponent`. In addition, you may want some children to be of different component then you can add something like this:
```javascript
//Assume that card is imported already.
const dummyData = [
...,
{
    child: Card,
    props: {
      title: 'Frozen Yogurt',
      category: 'Coding',
      tags: ['some', 'random', 'tag'],
      //props that are expected by the Card component.
    },
  },
...,
```

krushiraj committed:
update 2

Worked on data filter to be as generic as possible.

Added a search bar for data items container.

krushiraj committed:
Fixed for mobile view

- Made required changes to make it responsive.

TODOs
- Move dashboard to header of logged in view
- Move all data to vuex state
- Fetching data from API endpoint.

pbteja1998 committed:
Merge branch 'master' into dev-issue-9

krushiraj committed:
Added state to store

pbteja1998 committed:
Merge branch 'master' into dev-issue-9
